api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java

```diff
@@ -210,22 +211,51 @@ public Encounter saveEncounter(Encounter encounter) throws APIException {
 				//The logic in saveObs evicts the old obs instance, so we need to update the collection
 				//with the newly loaded and voided instance, apparently reloading the encounter
 				//didn't do the tick
-				toRemove.add(o);
-				toAdd.add(os.getObs(o.getId()));
-				toAdd.add(newObs);
+				obsToRemove.add(o);
+				obsToAdd.add(os.getObs(o.getId()));
+				obsToAdd.add(newObs);
 			}
 		}
-		
-		for (Obs o : toRemove) {
+
+		removeGivenObsAndTheirGroupMembersFromEncounter(obsToRemove, encounter);
+		addGivenObsAndTheirGroupMembersToEncounter(obsToAdd, encounter);
+		return encounter;
+	}
+
+	/**
+	 * This method will remove given list of obs and their group members from encounter
+	 *
+	 * @param obsToRemove list of obs that need to be removed recursively
+	 * @param encounter the encounter from which the obs will be removed
+	 */
+	private void removeGivenObsAndTheirGroupMembersFromEncounter(List<Obs> obsToRemove, Encounter encounter) {
+		for (Obs o : obsToRemove) {
 			encounter.removeObs(o);
+			Set<Obs> groupMembers = o.getGroupMembers(true);
+			if (CollectionUtils.isNotEmpty(groupMembers)) {
+				ArrayList<Obs> membersToRemove = new ArrayList<>(groupMembers);
```

Comment:
Well, here it is.
In `EncounterServiceImpl#saveEncounter():201` :-  
- We save the existing obs and add to a list named `toRemove`.
- We get a new obs after saving the existing one we add it to a list named `toAdd`.
- Other than this we get the existing obs from the database and add it to a list named `toAdd`.

Since I am removing/adding these obs recursively, I am passing the list and the encounter as a parameter to remove/add method. These are two private methods introduced which is supposed to iterate over list of obs and remove/add from encounter recursively. For recursion I need to do the same for all the group members of each obs in that list, in an iterator, I am calling `obs.getGroupMembers` which I need to pass as other members to add/remove. Since this call returns a `set` and the method accepts a `list`, I am wrapping it into a list and passing it.

Hope that answers your question.